### DISCUSSION TITLE
encodeGeometry

### DISCUSSION THREAD

**Discussion by RyugaRyuzaki** [zeux/meshoptimizer#667] [General]
I have a gemetry, anyone has any idea for this 👍 
``` private encodeGeometry(combine: THREE.BufferGeometry) {
    const {position, normal, ID} = combine.attributes;
    const {index} = combine;
    if (!position || !ID || !normal || !index) throw new Error('Missing value!');
    const {reorderMesh, encodeFilterOct, encodeIndexBuffer, encodeVertexBuffer} = MeshoptEncoder;
    // Step 1: reorderedMesh
    const reorderedMesh = reorderMesh(new Uint32Array(index.array), true, false); // true reOrder triangle mesh, true optimal size

    // Step 2: quantize position and normal
    const quantizedPosition = encodeFilterOct(position.array as Float32Array, position.array.length / 3, 4, 8); //  encodeFilterOct for position
    const quantizedNormal = encodeFilterOct(normal.array as Float32Array, normal.array.length / 3, 4, 8); //  encodeFilterOct for normal
    // Step 3: Compress data
    // Compress index
    const encodedIndex = encodeIndexBuffer(new Uint8Array(reorderedMesh[0]), reorderedMesh[1], 4); // encodeIndexBuffer for index

    // Compress position and normal
    const encodedPosition = encodeVertexBuffer(quantizedPosition, quantizedPosition.length, 4); //encodeVertexBuffer for position
    const encodedNormal = encodeVertexBuffer(quantizedNormal, quantizedNormal.length, 4); //encodeVertexBuffer for normal

    const newPosition = new THREE.InterleavedBuffer(encodedPosition, 4);
    const newNormal = new THREE.InterleavedBuffer(encodedNormal, 4);

    const newGeometry = new THREE.BufferGeometry();

    newGeometry.setAttribute('position', new THREE.InterleavedBufferAttribute(newPosition, 3, 0));
    newGeometry.setAttribute('normal', new THREE.InterleavedBufferAttribute(newNormal, 3, 0));
    newGeometry.setIndex(Array.from(encodedIndex));
    // transfer expressID
    const IDs = new Float32Array(ID.array);
    newGeometry.setAttribute('ID', new THREE.BufferAttribute(IDs, 1));

    combine.dispose();
    return newGeometry;
  }```

**Comment by donmccurdy** [comment]
@RyugaRyuzaki Usually Meshoptizer's compression is applied to an already-quantized mesh, downloaded to the client compressed, then decompressed before rendering. It looks like you're trying to both filter/compress and then render on the client device in one step, but the compressed data cannot be rendered without being decompressed first, so that may be a mistake. The geometry remains _quantized_ after decompression, and so uses 25-50% the size of default float32 geometry in three.js. Meshoptimizer assumes the user quantizes mesh data themselves.

If you'd like help with a three.js / meshoptimizer integration, the three.js forums (https://discourse.threejs.org/) would be the best place to ask. [This thread](https://github.com/mrdoob/three.js/issues/21492#issuecomment-1876151441) does give an example of using three.js + glTF Transform + Meshoptimizer to optimize a glTF model exported with THREE.GLTFExporter, but as mentioned above, that's not something you'd want to do immediately before rendering the model.

**Reply by RyugaRyuzaki** [reply]
Ah Sorry, i got this. thank you

**Comment by RyugaRyuzaki** [comment]
> @RyugaRyuzaki Usually Meshoptizer's compression is applied to an already-quantized mesh, downloaded to the client compressed, then decompressed before rendering. It looks like you're trying to both filter/compress and then render on the client device in one step, but the compressed data cannot be rendered without being decompressed first, so that may be a mistake. The geometry remains _quantized_ after decompression, and so uses 25-50% the size of default float32 geometry in three.js. Meshoptimizer assumes the user quantizes mesh data themselves.
> 
> If you'd like help with a three.js / meshoptimizer integration, the three.js forums (https://discourse.threejs.org/) would be the best place to ask. [This thread](https://github.com/mrdoob/three.js/issues/21492#issuecomment-1876151441) does give an example of using three.js + glTF Transform + Meshoptimizer to optimize a glTF model exported with THREE.GLTFExporter, but as mentioned above, that's not something you'd want to do immediately before rendering the model.
hi @donmccurdy thank for quickly reply. 
```  
scene.add(mesh);
const glb = (await exporter.parseAsync(scene, this.opts)) as ArrayBuffer;```
 How can i compress this without gltfpack.exe?
